docs/contributors/coding-guidelines.md

```diff
@@ -201,6 +201,246 @@ alert( 'My name is ' + name + '.' );
+
+```js
+/**
+ * Renders a toolbar.
+ *
+ * @param {Object} props             Component props.
+ * @param {string} [props.className] Class to set on the container `<div />`.
+ */
+```
+
+When a function does not include a `return` statement, it is said to have a `void` return value. It is not necessary to include a `@return` tag if the return type is `void`.
+
+If a function has multiple code paths where some (but not all) conditions result in a `return` statement, you can document this as a union type including the `void` type.
+
+```js
+/**
+ * Returns a configuration value for a given key, if exists.
+ * 
+ * @param {string} key Configuration key to retrieve.
+ * 
+ * @return {*|void} Configuration value, if exists.
+ */
+function getConfigurationValue( key ) {
+	if ( config.hasOwnProperty( key ) ) {
+		return config[ key ];
+	}
+}
+```
+
+When documenting a [function type](https://github.com/WordPress/gutenberg/blob/add/typescript-jsdoc-guidelines/docs/contributors/coding-guidelines.md#record-types), you must always include the `void` return value type, as otherwise the function is inferred to return a mixed ("any") value, not a void result.
+
+```js
+/**
+ * An apiFetch middleware handler. Passed the fetch options, the middleware is
+ * expected to call the `next` middleware once it has completed its handling.
+ *
+ * @typedef {(options:WPAPIFetchOptions,next:WPAPIFetchMiddleware)=>void} WPAPIFetchMiddleware
+ */
+```
+
+### Documenting Examples
+
+Because the documentation generated using the `@wordpress/docgen` tool will include `@example` tags if they are defined, it is considered a best practice to include usage examples for functions and components. This is especially important for documented members of a package's public API.
+
+When documenting an example, use the markdown <code>\`\`\`</code> code block to demarcate the beginning and end of the code sample. An example can span multiple lines.
+
+```js
+/**
+ * Given the name of a registered store, returns an object of the store's
+ * selectors. The selector functions are been pre-bound to pass the current
+ * state automatically. As a consumer, you need only pass arguments of the
+ * selector, if applicable.
+ *
+ * @param {string} name Store name.
+ *
+ * @example
+ * ```js
+ * select( 'my-shop' ).getPrice( 'hammer' );
+ * ```
+ *
+ * @return {Object} Object containing the store's selectors.
```

Comment:
Nit: It goes again the previous statement to avoid using `{Object}` :)

> When documenting a generic type such as `Object`, `Function`, `Promise`, etc., always include details about the expected record types.

Reply:
> Nit: It goes again the previous statement to avoid using `{Object}` :)

Good catch. I found myself needing to fix a few of the other examples I had pulled from real code 😄 In fact, with this one, I also already changed to remove the [reference to `wp.data.select` global](https://github.com/WordPress/gutenberg/blob/da0b0016869721b4844f8bd1a11e67aaae8e165d/packages/data/src/index.js#L83), since I'm not sure if this is really the direction we want to be promoting in examples for referencing the functions externally.

Reply:
> > Nit: It goes again the previous statement to avoid using `{Object}` :)
> 
> Good catch. I found myself needing to fix a few of the other examples I had pulled from real code 😄 In fact, with this one, I also already changed to remove the [reference to `wp.data.select` global](https://github.com/WordPress/gutenberg/blob/da0b0016869721b4844f8bd1a11e67aaae8e165d/packages/data/src/index.js#L83), since I'm not sure if this is really the direction we want to be promoting in examples for referencing the functions externally.

Corrected this and another error in b3c370026.